sys/linux/bpf_lsm.txt

```diff
@@ -0,0 +1,73 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/linux/bpf.h>
+
+resource fd_bpf_lsm_prog[fd]
+resource btf_id[int32]
+
+bpf_lsm_func_names = "bpf_lsm_path_mkdir"
+
+syz_btf_id_by_name(name ptr[in, string[bpf_lsm_func_names]]) btf_id
+
+type bpf_insn_tmpl[CODE, DST_REG, SRC_REG, OFF, IMM] {
+	code	const[CODE, int8]
+	dst	const[DST_REG, int8:4]
+	src	const[SRC_REG, int8:4]
+	off	const[OFF, int16]
+	imm	const[IMM, int32]
+}
+
+# Test program 0:
+# int BPF_PROG(test, ..., int ret)
+# {
+#	return 0;
+# }
+# 0: b7 00 00 00 00 00 00 00 r0 = 0
+# 1: 95 00 00 00 00 00 00 00 exit
+
+define insn_cnt_test0	2
+define code_test0_insn0	BPF_ALU64 | BPF_K | BPF_MOV
+define code_test0_insn1	BPF_JMP | BPF_K | BPF_EXIT
+
+insns_test0 {
+	insn0	bpf_insn_tmpl[code_test0_insn0, BPF_REG_0, 0, 0, 0]
+	insn1	bpf_insn_tmpl[code_test0_insn1, 0, 0, 0, 0]
+}
+
+type bpf_attr_bpf_lsm_test_tmpl[INSN_CNT, INSNS] {
```

Comment:
We need to reuse this struct description and instructions from bpf.txt. Maybe add some templates there if we need a specialization with a particular prog type.

Reply:
I assume I will need to change a lot of structs in bpf.txt to type templates to support replacing the type to constants and so we can fill in some pre-generated program. Not really sure that will be something we want?

Reply:
I looked more carefully at this file. It's not exactly how it needs to be done.
We don't need to do anything special related to tests/testing in descriptions. They need to describe kernel interface in all generality. Also the things you specify as template parameters, should not be template parameters.
E.g. `insn_cnt` should always be `bytesize8[insns, int32]` as we have it in the bpf.txt.
INSNS shouldn't be a template parameter as well. insns should be an array of generic instructions, as we have in bpf.txt.
Then in a unit test you specify, say, 3 concrete instructions and insn_cnt will be filled in automatically (if you use `AUTO`).
This should not be in descriptions as well:
```
insn0	bpf_insn_tmpl[code_test0_insn0, BPF_REG_0, 0, 0, 0]
```
instead in the test for the insns array you write something like: `[{0xabcd, 0x11, 0x0, 0x0, 0x0, 0x0}, ...]` (or whatever are the values for these code_test0_insn0 and BPF_REG_0).

Of the remaining template args you listed here:
```
bpf_prog_tmpl[TYPE, NINSN, INSNS, FLAGS, PROG_IFINDEX, EXPECTED_ATTACH_TYPE, BTF_FD, ...,  ATTACH_PROG_FD]
```
I am not sure FLAGS should be a template parameter. Can we the common flags type in bpf.txt?
Do we need PROG_IFINDEX? Does it need to be 0 for BPF_PROG_TYPE_LSM?

Refining the remaining fields (e.g. prog type and attach type) seems to be useful for all prog types. And we definitely don't want 30 copies of this struct.
Maybe something along the following lines (which will provide refined arguments for all prog types):
```
type bpf_prog_t[TYPE, ATTACH_TYPE, ... maybe few other parameters... ] {
	type			const[TYPE, int32]
...
}

bpf_prog [
  BPF_PROG_TYPE_SOCKET_FILTER        bpf_prog_t[BPF_PROG_TYPE_SOCKET_FILTER, ...]
  BPF_PROG_TYPE_KPROBE        bpf_prog_t[BPF_PROG_TYPE_KPROBE, ...]
  ...
  BPF_PROG_TYPE_LSM    bpf_prog_t[BPF_PROG_TYPE_LSM, ...]
]
```

Reply:
Take a look at:
https://github.com/google/syzkaller/blob/master/sys/linux/test/wireguard
The descriptions do not have anything related to this test. It's just the test fills in the generic data in a particular way to achieve the necessary behavior. We need the same here.

Reply:
Thank you for the detailed response. Let me make sure I understand correctly. Using what you suggest, I will come up with something like list
```
type bpf_prog_t[TYPE, ATTACH_TYPE, ... ] {
	type		const[TYPE, int32]
	ninsn		bytesize8[insns, int32]
	insns		ptr64[in, bpf_instructions]
...
}
```
And then,`insns`, the instructions of the bpf program, will be **generate dynamically by the fuzzer instead of using predefined values**. That is why we don't need to include `INSNS` in the type parameters.

However, what I though is that in the email where we discuss with KP (the author of BPF_LSM module), we have a conclusion that we should use a **predefined corpus of bpf lsm programs** for fuzzing different lsm hooks. I think there are advantages in both approach. It will integrate with Syzkaller better if we let the fuzzer generate the program. Yet, given that bpf program are verified before it can be attach to the hook, it might be hard to pass the verification stage if we are not using some pre-generated programs, and we will not be able to test the attachment stage.

In short, I specified the instructions inside the descriptions file not because I want to use it to test my descriptions (which I understand I could just specify in the test file), but because I want to have a predefined program that will pass the bpf verification and can be used by the fuzzer. Does this make sense? Thanks.

CC: @jedav

Reply:
If we just want to attach fixed programs to hooks, this should be done outside of the fuzzer. Like e.g. SELinux policy loaded by init. There is no work for a fuzzer here and no benefit of putting this into descriptions then if we don't use any fuzzer randomness.
If it's hard for the fuzzer to generate programs that pass validation (is it?), we should help it. And there are several ways to achieve this. We can either add more specialized instruction templates like these ones:
https://github.com/google/syzkaller/blob/master/sys/linux/bpf.txt#L320-L343
and/or programs templates:
https://github.com/google/syzkaller/blob/master/sys/linux/bpf.txt#L213-L217
and/or we could pre-seed the corpus with these fixed programs, search here for "use these as seeds for corpus":
https://groups.google.com/d/msg/syzkaller/XwJCzqr-OG8/1F1I__lfCAAJ

@sinkap

Reply:
Hm, I thought we wanted to choose at random from a set of fixed programs (currently one, to be expanded) and random hook-points.

Reply:
But why artificially restrict ourselves and limit the power of the fuzzer?
How will these fixed programs look like? Trivial ones (like "return 0") fuzzer should be able to generate given good descriptions.

Reply:
I changed it so that the fuzzer can generate arbitrary program now.